File: C++/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
 public:
  ListNode *removeNthFromEnd(ListNode *head, int n) {
    ListNode *ptr;
    ptr = head;
    int total = 0;
    while (ptr != NULL) {
      ptr = ptr->next;
      total++;
    }
    int delete_index = total - n - 1;
    if (delete_index < 0) {
      head = head->next;
      return head;
    }
    ptr = head;
    while (delete_index--) {
      ptr = ptr->next;
    }
    if ((ptr->next)->next) {
      ptr->next = (ptr->next)->next;
    } else {
      ptr->next = NULL;
    }
    return head;
  }
};

File: Python/0019-remove-nth-node-from-end-of-list.py
# time complexity: O(n)
# space complexity: O(1)
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        total = 0
        countNode = resNode = head
        while countNode:
            total += 1
            countNode = countNode.next
        delIdx = total - n - 1
        if delIdx < 0:
            head = head.next
            return head
        currentIdx = 0
        while resNode:
            if currentIdx == delIdx:
                if resNode.next.next:
                    resNode.next = resNode.next.next
                else:
                    resNode.next = None
                return head
            else:
                resNode = resNode.next
            currentIdx += 1
        return head


root = ListNode(1)
root.next = ListNode(2)
root.next.next = ListNode(3)
root.next.next.next = ListNode(4)
root.next.next.next.next = ListNode(5)
print(Solution().removeNthFromEnd(root, 2))
